x/stakeibc/keeper/msg_server_lsm_liquid_stake.go

```diff
@@ -2,21 +2,230 @@ package keeper
+	transfertypes "github.com/cosmos/ibc-go/v5/modules/apps/transfer/types"
+	"github.com/golang/protobuf/proto" //nolint:staticcheck
 
+	stakingtypes "github.com/cosmos/cosmos-sdk/x/staking/types"
+
+	icacallbackstypes "github.com/Stride-Labs/stride/v8/x/icacallbacks/types"
+	icqtypes "github.com/Stride-Labs/stride/v8/x/interchainquery/types"
+
+	recordskeeper "github.com/Stride-Labs/stride/v8/x/records/keeper"
 	"github.com/Stride-Labs/stride/v8/x/stakeibc/types"
 )
 
+// Exchanges a user's LSM tokenized shares for stTokens using the current redemption rate
+// The LSM tokens must live on Stride with an IBC denomination before this function is called
+//
+// The typical flow:
+//   - A staker tokenizes their delegation on the host zone
+//   - The staker IBC transfers their tokenized shares to Stride
+//   - They then call LSMLiquidStake
+//   - The staker's LSM Tokens are sent to the Stride module account
+//   - The staker recieves stTokens
+//
+// As a safety measure, at period checkpoints, the validator's exchange rate is queried and the transaction
+// is not settled until the query returns
+// As a result, this transaction has been split up into a (1) Start and (2) Finish function
+//   If no query is needed, (2) is called immediately after (1)
+//   If a query is needed, (2) is called in the query callback
+//
+// The transaction response indicates if the query occurred by returning an attribute `TransactionComplete` set to false
 func (k msgServer) LSMLiquidStake(goCtx context.Context, msg *types.MsgLSMLiquidStake) (*types.MsgLSMLiquidStakeResponse, error) {
 	ctx := sdk.UnwrapSDKContext(goCtx)
-	_ = ctx
-	// TODO [LSM]
 
-	return &types.MsgLSMLiquidStakeResponse{}, nil
+	lsmLiquidStake, err := k.StartLSMLiquidStake(ctx, msg)
+	if err != nil {
+		return nil, err
+	}
+
+	if k.ShouldQueryValidatorExchangeRate(ctx, lsmLiquidStake.Validator, msg.Amount) {
+		if err := k.SubmitValidatorExchangeRateQuery(ctx, lsmLiquidStake); err != nil {
+			return nil, err
+		}
+		return &types.MsgLSMLiquidStakeResponse{TransactionComplete: false}, nil
+	}
+
+	if err := k.FinishLSMLiquidStake(ctx, lsmLiquidStake); err != nil {
+		return nil, err
+	}
+
+	return &types.MsgLSMLiquidStakeResponse{TransactionComplete: true}, nil
+}
+
+// StartLSMLiquidStake runs the transactional logic that occurs before the optional query
+// This includes validation on the LSM Token, and the escrowing of tokens
+func (k Keeper) StartLSMLiquidStake(ctx sdk.Context, msg *types.MsgLSMLiquidStake) (types.LSMLiquidStake, error) {
+	// Get the denom trace from the IBC hash - this includes the full path and base denom
+	denomTrace, err := k.GetLSMTokenDenomTrace(ctx, msg.LsmTokenIbcDenom)
+	if err != nil {
+		return types.LSMLiquidStake{}, err
+	}
+
+	// Get the host zone and validator address from the path and base denom respectively
+	lsmTokenBaseDenom := denomTrace.BaseDenom
+	hostZone, err := k.GetHostZoneFromLSMTokenPath(ctx, denomTrace.Path)
```

Comment:
yeah using the prefix feels a bit messy since every chain has a different prefix length. Plus it's the valoper prefix that we'd need to check (e.g. cosmosvaloper), and we don't have access to that value for each host (we only have the base address prefix).

But since we're limiting tokens to ibc denom's that were no greater than 1 hop away, I don't have any concern with using the channel-id